packages/common/src/api/tan-query/useFavoriteTrack.ts

```diff
@@ -0,0 +1,171 @@
+import { Id } from '@audius/sdk'
+import { useMutation, useQueryClient } from '@tanstack/react-query'
+import { useDispatch } from 'react-redux'
+
+import { useAudiusQueryContext } from '~/audius-query'
+import { useAppContext } from '~/context/appContext'
+import { Name } from '~/models/Analytics'
+import { Feature } from '~/models/ErrorReporting'
+import { ID } from '~/models/Identifiers'
+import { Track } from '~/models/Track'
+import { accountActions } from '~/store/account'
+import { tracksSocialActions } from '~/store/social'
+
+import { useCurrentUserId } from './useCurrentUserId'
+import { getTrackQueryKey } from './useTrack'
+import { useUser } from './useUser'
+import { primeTrackData } from './utils/primeTrackData'
+
+type FavoriteTrackArgs = {
+  trackId: ID
+  source?: string
+}
+
+export const useFavoriteTrack = () => {
+  const { audiusSdk, reportToSentry } = useAudiusQueryContext()
+  const queryClient = useQueryClient()
+  const dispatch = useDispatch()
+  const { data: currentUserId } = useCurrentUserId()
+  const { data: currentUser } = useUser(currentUserId)
+  const {
+    analytics: { track: trackEvent }
+  } = useAppContext()
+
+  return useMutation({
+    mutationFn: async ({ trackId }: FavoriteTrackArgs) => {
+      if (!currentUserId) throw new Error('User ID is required')
+      const sdk = await audiusSdk()
+      await sdk.tracks.favoriteTrack({
+        trackId: Id.parse(trackId),
+        userId: Id.parse(currentUserId)
+      })
+    },
+    onMutate: async ({ trackId, source }) => {
+      if (!currentUserId || !currentUser) {
+        // TODO: throw toast and redirect to sign in
+        throw new Error('User ID is required')
+      }
+
+      // Cancel any outgoing refetches
+      await queryClient.cancelQueries({ queryKey: getTrackQueryKey(trackId) })
+
+      // Snapshot the previous values
+      const previousTrack = queryClient.getQueryData<Track>(
+        getTrackQueryKey(trackId)
+      )
+      if (!previousTrack) throw new Error('Track not found')
+
+      // Don't allow favoriting your own track
+      if (previousTrack.owner_id === currentUserId)
```

Comment:
would we want to do this check in the mutationFn, or does onMutate run before?

Reply:
`onMutate` runs first and throwing in it will go straight to `onError` and prevent `mutaitonFn` from getting called at all